### PR TITLE
fix #62531 add negative text filtering to Problems

### LINKED ISSUE #62531: Support negative filters on text
It seems like attempting to filter out certain text from errors or warnings does not work. Take the following for example:

Filtering to include a warning:

Filtering to exclude the warning:

I have also tried using quotes. It seems like the exclude `!` flag does not apply to text.

### PR DESCRIPTION
**Pull request by gjsjohnmurray**
This PR fixes #62531 

If the first character of the filter field value is `!` then only problems that don't match the rest of the value will display.

To test:
1. Set up a situation where there are multiple problems listed and no filter is specified.
2. Enter a text filter that removes some of the problems.
3. Then prepend a `!` to your filter and confirm you see only the problems that the previous step had omitted.

### CODE REVIEW

**Comment by gjsjohnmurray** [comment]
@sandy081 please review this.

**Comment by gjsjohnmurray** [comment]
@sandy081 I think it'd be good to get this merged early in the life of 1.54 so it can get some miles on the clock 🙏

**Comment by sandy081** [comment]
Sorry for the delay. I will do first thing reviewing this tomorrow.

File: src/vs/workbench/contrib/markers/browser/markersTreeViewer.ts [hunk 1]
```diff
@@ -533,10 +533,16 @@ export class Filter implements ITreeFilter<MarkerElement, FilterData> {
 			return false;
 		}
 
-		const uriMatches = FilterOptions._filter(this.options.textFilter, basename(resourceMarkers.resource));
-
-		if (this.options.textFilter && uriMatches) {
-			return { visibility: true, data: { type: FilterDataType.ResourceMarkers, uriMatches } };
+		const negate = this.options.textFilter.startsWith('!');
+		const textFilter = this.options.textFilter.substring(negate ? 1 : 0);
```
Comment (sandy081) [inline src/vs/workbench/contrib/markers/browser/markersTreeViewer.ts:537]: Seems like textFilter & negate options are being computed outside and everywhere. Instead I would suggest to move these semantics into FilterOptions and change `textFilter` type to `{text: string, negate: boolean} | undefined`.

**Review by sandy081 — CHANGES_REQUESTED** [review]

File: src/vs/workbench/contrib/markers/browser/markersTreeViewer.ts (hunk 1 above)
Comment (gjsjohnmurray) [reply src/vs/workbench/contrib/markers/browser/markersTreeViewer.ts]: Good idea. I have pushed the changes.

File: src/vs/workbench/contrib/markers/browser/markersTreeViewer.ts [hunk 2]
```diff
@@ -564,41 +570,57 @@ export class Filter implements ITreeFilter<MarkerElement, FilterData> {
...
+		const uriMatches = FilterOptions._filter(text, basename(relatedInformation.raw.resource));
+		const messageMatches = FilterOptions._messageFilter(text, paths.basename(relatedInformation.raw.message));
 
-		if (uriMatches || messageMatches) {
-			return { visibility: true, data: { type: FilterDataType.RelatedInformation, uriMatches: uriMatches || [], messageMatches: messageMatches || [] } };
+		const anyMatched = (uriMatches || messageMatches) ? true : false;
+		if (anyMatched && negate) {
+			return false;
+		} else if (anyMatched || negate) {
+			return negate ? false : { visibility: true, data: { type: FilterDataType.RelatedInformation, uriMatches: uriMatches || [], messageMatches: messageMatches || [] } };
```
Comment (sandy081) [inline src/vs/workbench/contrib/markers/browser/markersTreeViewer.ts:623]: Is it an oversight here to return `false` when `negate`?

**Review by sandy081 — CHANGES_REQUESTED** [review]

File: src/vs/workbench/contrib/markers/browser/markersTreeViewer.ts (hunk 2 above)
Comment (sandy081) [reply src/vs/workbench/contrib/markers/browser/markersTreeViewer.ts:617]: @gjsjohnmurray Not sure if you have seen this comment?

**Review by sandy081 — APPROVED** [review]

**Comment by sandy081** [comment]
@gjsjohnmurray Thanks for the changes. Merging the PR.

**Comment by GitMensch** [comment]
That's one of the long awaited "minor issues" that comes up quite regularly. Thanks for having it done so far. Of course people will wonder about combined filters next `"include/" !"inc1.ext" !"inc2.ext"`(all problems in all includes but not inc1 and inc2)...
I suggest to either add/adjust that before the release, or have the documentation be explicit about that combinations are either "all applied" or "all suppressed"

**Comment by sandy081** [comment]
@GitMensch not sure what you mean by your filter `"include/" !"inc1.ext" !"inc2.ext"` - this negative filtering support is added only for text filters - for eg do not show problems with text `folder` - for resource we already support globs (inclusive and exclusive)

**Comment by GitMensch** [comment]
Sounds good. That remark was about a combined filtering in general:

* file: only in folders that end in `include`
* file: not files named "inc1.ext" or "inc2.ext"
* text "overflow"
* negated filter "possible"

(that was a help request from a co-worker some weeks ago to show all overflow warnings but filter out the ones that explicit said something about "possible overflow" [without recompiling everything])

**Comment by GitMensch** [comment]
When I see all problems in a huge workspace I sometimes conclude to not or only investigate two (or three or...) issues.

`problem1 problem2` shows no problem, `problem1, problem2`, `problem1 + problem2`, `problem1 || problem2` and `"problem1"` also show no issues, so the negation can't work either (as the common text filter it applies to the complete filter text).

Is "combined filtering" (which doesn't seems to be possible yet) in the pipeline? I did not found an issue for this with "filter pane" search.

**Pull request merged**